cEP-0016.md

```diff
@@ -0,0 +1,332 @@
+     or any other branch onto the pull request, message is displayed to perform
+     a `git rebase` instead.
+7. For revert commits:
+   - If the commit type from the output of `VCSCommitBear` indicates that the
+     commit is a `git revert` commit, inspection is done in `RevertCommitBear`
+     otherwise return.
+   - In `RevertCommitBear`, the patches of both commits are compared to
+     ensure that they are in fact reverse of each other. Minor differences are
+     allowed.
+   - In case of total mismatch, a note in the commit message highlighting that
+     it is not a clean revert is expected.
+   - A message suggesting improvements in the commit message is displayed.
+
+## Code Prototypes
+
+A prototype for identifying GitHub Pull Request temporary merge commits:
+
+```python
+    if len(shortlog) > shortlog_length:
+        github_pr_merge_commit_regex = re.compile(r'^Merge [0-9a-f]+ into [0-9a-f]+$|....')
+
+        match = re.fullmatch(github_pr_merge_commit_regex, shortlog)
+        if match:
+            # ignore this commit from any further inspection
+            # inspect the unmerged parent of this commit instead
+```
+
+A prototype of `VCSCommitBear` :
+
+```python
+    # import statements
+
+    class CommitResult(HiddenResult):
+        def __init__(self, origin, commit_info, message):
+            Result.__init__(self, origin, commit_info)
+            self.commit_info = commit_info
+            self.message = message
+
+
+    class VCSCommitBear(GlobalBear):
+        __metaclass__ = abc.ABCMeta
+        AUTHORS = {'The coala developers'}
+        AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+        LICENSE = 'AGPL-3.0'
+
+        @abc.abstractmethod
+        def get_head_commit(self):
+            """
+            Return the commit message from the head commit
+            """
+
+        def analyze_git_commit(self, stdout):
+            commit_info = dict()
+
+            commit_sha = run_shell_command('git rev-parse HEAD')[0].strip('\n')
+
+            commit_type = []
+
+            stdout = stdout.strip('\n')
+
+            if 'ci skip' in stdout or 'skip ci' in stdout:
+                commit_type.append('ci_skip_commit')
+
+            get_parent_commits = 'git log --pretty=%P -n 1 ' + commit_sha
+            parent_commits, _ = run_shell_command(get_parent_commits)
+            parent_commits_list = parent_commits.split('\n')
+
+            if len(parent_commits_list) > 2:
+                commit_type.append('merge_commit')
+
+            get_modified_files = 'git show --pretty="" --name-only ' + commit_sha
+            modified_files, _ = run_shell_command(get_modified_files)
+            modified_files_list = modified_files.split('\n')
+
+            commit_info['commit_sha'] = commit_sha
+            commit_info['commit_type'] = commit_type
+            commit_info['modified_files'] = modified_files_list
+
+            commit_info['commit_type'] = commit_type
+
+            return str(commit_info)
```

Comment:
this isnt correct

Reply:
Why ?

Reply:
because returning a stringified version of a string is not the best way to go, because it's neither machine-friendly nor it is human-friendly. Either you make a nice string format there to print the data in a human-friendly way, or return the dict, and for human eyes, you pass it through a presentation function (which will use a format string to nicely print it out), while the dict can be used for M2M stuff (like JSON conversion, or extending it like a lib).

Reply:
I have written a comment saying the returned string would be converted to dictionary using (JSON conversion) . If I try to return a dictionary, it gives error

Reply:
If this is immutable I would suggest use something like tuple or namedtuple. makes work alot easier. Well if JSON conversion is easier go for that. It totally depends on you.
https://github.com/coala/coala/blob/master/coalib/bearlib/languages/documentation/DocstyleDefinition.py#L18
https://docs.python.org/2/library/collections.html#collections.namedtuple